### PR TITLE
ci(matrix): #502 latest-only matrix default + dep-driven retained rebuild

### PR DESCRIPTION
**Pull request by oorabona**
Closes #502

## Summary

Every push to master previously rebuilt **all** retained versions of every container (e.g. `openresty` 1.29.2.5 + 1.29.2.4 + 1.29.2.3 × 2 archs = 6 jobs), wasting runner time and registry rate-limit budget. The retention list is meant for **dependency-driven rebuilds** (propagating a security patch in alpine/openssl to every supported line), not for every code push.

This PR ships the matrix-default change and the openresty/build `$VERSION` fix called out in #502.

## Matrix logic

- `helpers/variant-utils.sh:list_build_matrix` gains 3rd arg `include_all_retained` (default `false` → filters to `is_latest_version=true`).
- New `compute_expand_retained_map` helper computes per-container expand signal from git diff. Priority chain: diff-failed sentinel → unknown event → `pull_request` (full smoke) → `build_all_retained=true` override → any file in `<container>/...` changed → default false.
- `.github/actions/detect-containers/action.yaml` wires the map through to the per-container `list_container_builds` call. New `event_name` + `build_all_retained` inputs; new `expand_retained_map` output.
- `auto-build.yaml` gains `workflow_dispatch.inputs.build_all_retained` (mirrored in `workflow_call.inputs`) for operator override — CVE fan-out, resync after a registry incident.

## Postgres opt-out

`postgres/variants.yaml` declares `build.always_all_versions: true`. All 4 supported PG majors (15/16/17/18) rebuild on every push — they're separate long-lived release streams, not a rolling history. The flag wins over `include_all_retained=false` inside `list_build_matrix`.

## openresty/build $VERSION fix

`openresty/build:14` previously called `version.sh --upstream` unconditionally — wrong artifacts for any retained-version rebuild. Now honors `$VERSION` when set (strips `-alpine` suffix, fails loud on empty result), falls back to `version.sh --upstream` for local/manual invocation.

## Consumer / caller wiring

- `make:524` passes `"true"` 3rd arg so `./make list-builds` and `scripts/cleanup-outdated-tags.sh` get the full retained set. Without it, the monthly cleanup workflow would have purged retained tags.
- `recreate-manifests.yaml` + `validate-version-scripts.yaml` pass `build_all_retained: true`. Their `force_rebuild: true` intent requires full retained matrix; otherwise they would silently collapse to latest-only.
- `upstream-monitor.yaml` (lines 514, 1026) passes `-f build_all_retained=true` on its `gh workflow run auto-build.yaml` dispatches. `workflow_dispatch` carries no `before..after` diff context.
- `auto-build.yaml` path filters broadened from per-file allowlist to `*/**` plus denylist (`docs/**`, `examples/**`, `tests/**`, `test-harness/**`, `test-logs/**`, `archive/**`). The narrow allowlist had been brittle — new build-input file types (build hooks, prepare-build-context scripts, Dockerfile-COPYed assets) slipped past the trigger.

## detect-containers edge cases

- `diff_rc` initialized to 0 at find-containers entry so the sentinel-write step runs safely when `inputs.container` is set (single-container target path).
- Fail-loud guard in compute-expand-map: when the diff-failed sentinel is present AND no containers were detected, emit `::error::` and exit 1 instead of silently producing `map={}`.
- `scope_versions` being set bypasses the latest-only filter for that container's matrix expansion. Otherwise `scope_versions=1.29.2.4-alpine` targeting a retained version would return zero builds.

## Tests

- 15 new bats tests for `compute_expand_retained_map` (event types, build_all_retained string compare, exact-prefix match, sentinel handling, Dockerfile / variants.yaml / build trigger regression-locks).
- 10 new bats tests for `always_all_versions` + `list_build_matrix include_all_retained` (default latest-only, single-version container, postgres opt-out, stderr notice emission).
- 5 new bats tests for `openresty/build $VERSION` resolution + fallback + edge cases.
- 4 pre-existing tests adjusted to `run --separate-stderr` because the new observability `::notice::` on stderr was polluting jq output capture in the older tests.

All bats green (63 passing on this branch); `shellcheck` clean on modified shell.

## Post-merge expectations

- `openresty` matrix on master push: **6 jobs → 2 jobs** (3 versions × 2 archs → latest × 2 archs).
- `postgres` matrix: unchanged (all 4 majors always build via opt-out).
- `upstream-monitor` dep-bump flow: unchanged (still propagates to all retained versions via explicit `build_all_retained=true` dispatch).
- PR smoke builds: unchanged (full retained-version matrix preserved).
- Operator override available via `workflow_dispatch` form: `build_all_retained=true`.

## Out of scope (deferred)

- Investigation of the 0.1s fail-fast on `openresty 1.29.2.5-alpine` build (#502 Fix C) — likely self-resolves once the matrix stops compounding the flake noise; will observe master after merge and file a follow-up if the build still fails.
- Architectural inversion of `list_container_builds`'s 3rd arg default (full output → opt-in filter) — recorded as follow-up to prevent future "consumer forgot the arg → silent data loss" regressions.

## Test plan

- [ ] CI passes (shellcheck, bats, validate-version-scripts)
- [ ] First post-merge master `auto-build` run shows openresty matrix = 2 jobs (latest × 2 archs), not 6
- [ ] First post-merge master `auto-build` run shows postgres matrix unchanged (all 4 majors)
- [ ] Manual `workflow_dispatch` with `build_all_retained=true` produces full retained matrix
- [ ] Next upstream-monitor dep bump correctly rebuilds all retained versions of the affected container

### CODE REVIEW

File: .github/workflows/auto-build.yaml [hunk 1]
```diff
@@ -53,52 +53,53 @@ on:
...
+      - '*/**'
+      - 'make'                                  # Top-level build orchestrator
+      - '.github/actions/build-container/**'    # Composite action used by build matrix
+      - '.github/actions/dockerhub-mirror/**'   # Composite action used by build pipeline
+      - '!docs/**'                              # Jekyll site + ADRs (no build impact)
+      - '!examples/**'                          # Docker Compose usage examples
+      - '!tests/**'                             # Test suite (has its own CI)
+      - '!test-harness/**'                      # Test fixtures
+      - '!test-logs/**'                         # Generated test logs
+      - '!archive/**'                           # Archived containers
```
[the remark this answers is not in the document]
Comment (oorabona) [reply .github/workflows/auto-build.yaml:82]: Acknowledged. Per minimatch defaults used by GitHub Actions (`dot: false`), the `*` glob does not match leading-dot path segments, so `*/**` does not match `.github/**`. Adding an explicit `!.github/**` denylist would conflict with the positive `.github/actions/build-container/**` and `.github/actions/dockerhub-mirror/**` entries above (negative patterns subtract from the match set regardless of subsequent positives). The denylist was tightened in a follow-up commit to exclude `**/*.md`, `**/README*`, `**/LICENSE*`, `**/CHANGELOG*`, `**/docs/**`, `**/tests/**` (with `**/LAST_REBUILD.md` re-included as the explicit upstream-monitor rebuild trigger) — this covers the doc-only-change case raised here without the `.github/` collision.
